lib/spack/llnl/util/filesystem.py

```diff
@@ -510,8 +598,145 @@ def basenames(self):
 
         >>> l = LibraryList(['/dir1/liba.a', '/dir2/libb.a', '/dir3/liba.a'])
         >>> assert l.basenames == ['liba.a', 'libb.a']
+        >>> h = HeaderList(['/dir1/a.h', '/dir2/b.h', '/dir3/a.h'])
+        >>> assert h.basenames == ['a.h', 'b.h']
+        """
+        return list(dedupe(os.path.basename(x) for x in self.files))
+
+    @property
+    def names(self):
+        """Stable de-duplication of file names in the list
+
+        >>> h = HeaderList(['/dir1/a.h', '/dir2/b.h', '/dir3/a.h'])
+        >>> assert h.names == ['a', 'b']
+        """
+        return list(dedupe(x.split('.')[0] for x in self.basenames))
+
+    def __getitem__(self, item):
+        cls = type(self)
+        if isinstance(item, numbers.Integral):
+            return self.files[item]
+        return cls(self.files[item])
+
+    def __add__(self, other):
+        return self.__class__(dedupe(self.files + list(other)))
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __eq__(self, other):
+        return self.files == other.files
+
+    def __len__(self):
+        return len(self.files)
+
+    def joined(self, separator=' '):
+        return separator.join(self.files)
+
+    def __repr__(self):
+        return self.__class__.__name__ + '(' + repr(self.files) + ')'
+
+    def __str__(self):
+        return self.joined()
+
+
+class HeaderList(FileList):
+    """Sequence of absolute paths to headers
+
+    Provides a few convenience methods to manipulate header paths and get
+    commonly used compiler flags or names
+    """
+
+    def __init__(self, files):
+        super(HeaderList, self).__init__(files)
+
+        self._macro_definitions = []
+
+    @property
+    def headers(self):
+        return self.files
+
+    @property
+    def include_flags(self):
+        """Include flags
+
+        >>> h = HeaderList(['/dir1/a.h', '/dir1/b.h', '/dir2/c.h'])
+        >>> assert h.cpp_flags == '-I/dir1 -I/dir2'
         """
-        return list(dedupe(os.path.basename(x) for x in self.libraries))
+        return ' '.join(['-I' + x for x in self.directories])
+
+    @property
+    def macro_definitions(self):
+        """Macro definitions
+
+        >>> h = HeaderList(['/dir1/a.h', '/dir1/b.h', '/dir2/c.h'])
+        >>> h.add_macro('-DBOOST_LIB_NAME=boost_regex')
+        >>> h.add_macro('-DBOOST_DYN_LINK')
+        >>> assert h.macro_definitions == '-DBOOST_LIB_NAME=boost_regex -DBOOST_DYN_LINK'  # noqa
+        """
+        return ' '.join(self._macro_definitions)
+
+    @property
+    def cpp_flags(self):
+        """Include flags + macro definitions
+
+        >>> h = HeaderList(['/dir1/a.h', '/dir1/b.h', '/dir2/c.h'])
+        >>> h.add_macro('-DBOOST_DYN_LINK')
+        >>> assert h.macro_definitions == '-I/dir1 -I/dir2 -DBOOST_DYN_LINK'
+        """
+        return self.include_flags + ' ' + self.macro_definitions
+
+    def add_macro(self, macro):
+        """Add a macro definition"""
+        self._macro_definitions.append(macro)
```

Comment:
@alalazo Does this satisfy your request from https://github.com/LLNL/spack/pull/3377#issuecomment-290334034?